### DISCUSSION TITLE
ConnectionRecoveryError event handler on ConnectionFactory

### DISCUSSION THREAD

**Discussion by ChristophBachmann** [Havret/dotnet-activemq-artemis-client#305] [Q&A] [answered]
Hello there,

I have defined typed Consumers as described here: https://havret.io/activemq-artemis-net-core

Now I would like to subscribe to the IConnection.ConnectionRecoveryError event, which is only possible if I have an IConnection object, which I do not have. I have the only possibility to call "ConfigureConnectionFactory" which allows me to set the RecoveryPolicy e.g. to "RecoveryPolicyFactory.ConstantBackoff" - which is nice - but I would also like to subscribe to the ConnectionRecoveryError here as I then would like to send an email or something like this ("Connection recovery failed" what ever).
![2021-11-22 13_26_55-Window](https://user-images.githubusercontent.com/57751050/142861839-6306825e-5bc0-4d04-92ce-3d0e6471a715.png)

How can I do that? Or would it be an idea, to have the ConnectionRecoveryError event on the ConnectionFactory?

Thanks and best regards
Christoph

**Comment by Havret** [comment, marked as answer]
Hi @ChristophBachmann,

You're right. Currently, there is no way to subscribe to ConnectionRecoveryError easily. In most of my use cases, I'm relying on some custom health-check to detect connectivity issues.

I think the easiest way to make this available is by extending DI API with an additional method that allows the client to configure a connection.

It may look as follows: 

```csharp
public void ConfigureServices(IServiceCollection services)
{
    services.AddActiveMq(name: "my-artemis-cluster", endpoints: new[] { Endpoint.Create(host: "localhost", port: 5672, "artemis", "artemis") })
            .ConfigureConnection((serviceProvider, connection) =>
            {
                connection.ConnectionClosed += (_, args) =>
                {
                    Console.WriteLine($"Connection closed: ClosedByPeer={args.ClosedByPeer}, Error={args.Error}");
                };
                connection.ConnectionRecovered += (_, args) =>
                {
                    Console.WriteLine($"Connection recovered: Endpoint={args.Endpoint}");
                };
                connection.ConnectionRecoveryError += (_, args) =>
                {
                    Console.WriteLine($"Connection recovered error: Exception={args.Exception}");
                };
            })
            .AddConsumer("a1", RoutingType.Multicast, "q1", async (message, consumer, token, serviceProvider) =>
            {
                Console.WriteLine("q1: " + message.GetBody<string>());
                await consumer.AcceptAsync(message);
            })
            .AddConsumer("a1", RoutingType.Multicast, "q2", async (message, consumer, token, serviceProvider) =>
            {
                Console.WriteLine("q2: " + message.GetBody<string>());
                await consumer.AcceptAsync(message);
            })
            .AddProducer<MyTypedMessageProducer>("a1", RoutingType.Multicast)
            .EnableQueueDeclaration()
            .EnableAddressDeclaration();

    services.AddActiveMqHostedService();
}
```

If you're happy with this API, the PR is ready to be merged --> https://github.com/Havret/dotnet-activemq-artemis-client/pull/308

Best,
Havret

**Reply by ChristophBachmann** [reply]
Hi @Havret 
Wow, fast answer - thank you so much! Yes I'm really happy with that!
Best,
Christoph

**Reply by Havret** [reply]
@ChristophBachmann version 2.4.0 is available on NuGet.
